ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryExecutor.java

```diff
@@ -261,30 +274,70 @@ PersistentQueryMetadata buildPersistentQuery(
         .map(userErrorClassifiers::and)
         .orElse(userErrorClassifiers);
 
-    return new PersistentQueryMetadataImpl(
-        statementText,
-        querySchema,
-        sources,
-        sinkDataSource,
-        planSummary,
-        queryId,
-        materializationProviderBuilder,
-        applicationId,
-        topology,
+    final boolean useBinPacked = true;
+    if (useBinPacked) {
+      BinPackedStreamsMetadata binPackedStreamsMetadata = getStream(sources);
+      BinPackedPersistentQueryMetadataImpl binPackedPersistentQueryMetadata = new BinPackedPersistentQueryMetadataImpl(
+          statementText,
+          querySchema.logicalSchema(),
+          querySchema,
+          sources,
+          planSummary,
+          applicationId,
+          topology,
+          binPackedStreamsMetadata,
+          runtimeBuildContext.getSchemas(),
+          config.getOverrides(),
+          queryId,
+          materializationProviderBuilder,
+          physicalPlan,
+          getUncaughtExceptionProcessingLogger(queryId),
+          sinkDataSource,
+          listener
+      );
+      binPackedStreamsMetadata.addQuery(classifier, streamsProperties, binPackedPersistentQueryMetadata, queryId);
+      return binPackedPersistentQueryMetadata;
+    } else {
+
+      return new PersistentQueryMetadataImpl(
+          statementText,
+          querySchema,
+          sources,
+          sinkDataSource,
+          planSummary,
+          queryId,
+          materializationProviderBuilder,
+          applicationId,
+          topology,
+          kafkaStreamsBuilder,
+          runtimeBuildContext.getSchemas(),
+          streamsProperties,
+          config.getOverrides(),
+          ksqlConfig.getLong(KSQL_SHUTDOWN_TIMEOUT_MS_CONFIG),
+          classifier,
+          physicalPlan,
+          ksqlConfig.getInt(KsqlConfig.KSQL_QUERY_ERROR_MAX_QUEUE_SIZE),
+          getUncaughtExceptionProcessingLogger(queryId),
+          ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_INITIAL_MS),
+          ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS),
+          listener,
+          scalablePushRegistry
+      );
+    }
+  }
+
+  private BinPackedStreamsMetadata getStream(final Set<SourceName> sources) {
+    for (BinPackedStreamsMetadata binPackedStreamsMetadata: streams) {
```

Comment:
I don't think we should be computing this here. It's a little brittle - if we change the selection algorithm or change the data structure we use to store the runtimes so that the traversal is in the wrong order we might pick different runtimes across restarts. Instead, I think we should compute the target runtime when building `QueryPlan` (see `EngineExecutor.plan`) and then we can store its name in the final plan that gets serialized to the command topic.